File: src/main/java/rip/hippo/mosey/asm/wrapper/MethodWrapper.java
/*
 * Mosey is a free and open source java bytecode obfuscator.
 *     Copyright (C) 2020  Hippo
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package rip.hippo.mosey.asm.wrapper;

import org.objectweb.asm.tree.*;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

/**
 * @author Hippo
 * @version 1.0.0, 7/1/20
 * @since 1.0.0
 */
public final class MethodWrapper {

    private final MethodNode methodNode;

    public MethodWrapper(MethodNode methodNode) {
        this.methodNode = methodNode;
    }

    public void addModifier(int modifier) {
        methodNode.access |= modifier;
    }

    public void removeModifier(int modifier) {
        methodNode.access &= ~modifier;
    }

    public void addVisibleAnnotation(AnnotationNode annotationNode) {
        if (methodNode.visibleAnnotations == null) {
            methodNode.visibleAnnotations = new ArrayList<>(1);
        }
        methodNode.visibleAnnotations.add(annotationNode);
    }

    public boolean hasModifier(int modifier) {
        return (methodNode.access & modifier) != 0;
    }

    public InsnList getInstructions() {
        return methodNode.instructions;
    }

    public List<TryCatchBlockNode> getTryCatchBlocks() {
        return methodNode.tryCatchBlocks;
    }

    public String getName() {
        return methodNode.name;
    }
}

File: src/main/scala/rip/hippo/mosey/asm/wrapper/MethodWrapper.scala
package rip.hippo.mosey.asm.wrapper

import java.util

import org.objectweb.asm.Attribute
import org.objectweb.asm.tree.{AnnotationNode, InsnList, MethodNode, TryCatchBlockNode}

/**
 * @author Hippo
 * @version 1.0.0, 7/8/20
 * @since 1.0.0
 */
final class MethodWrapper(methodNode: MethodNode) {

  def addVisibleAnnotation(annotationNode: AnnotationNode): Unit = {
    if (methodNode.visibleAnnotations == null) {
      methodNode.visibleAnnotations = new util.ArrayList()
    }
    methodNode.visibleAnnotations.add(annotationNode)
  }

  def addModifier(modifier: Int): Unit = methodNode.access |= modifier

  def removeModifier(modifier: Int): Unit = methodNode.access &= ~modifier

  def hasModifier(modifier: Int): Boolean = (methodNode.access & modifier) != 0

  def getInstructions: InsnList = methodNode.instructions

  def getTryCatchBlocks: util.List[TryCatchBlockNode] = methodNode.tryCatchBlocks

  def getAttributes: util.List[Attribute] = methodNode.attrs

  def createAttributes: Unit = methodNode.attrs = new util.ArrayList[Attribute]()

  def getMaxLocals: Int = methodNode.maxLocals

  def getName: String = methodNode.name
}
